tensorflow_quantum/core/ops/math_ops/inner_product_op.py

```diff
@@ -82,8 +161,60 @@ def inner_product(programs, symbol_names, symbol_values, other_programs):
         `tf.Tensor` with shape [batch_size, n_others] where `out[i][j]` is equal
             to the inner product of `programs[i]` with `symbol_values[i]`
             resolved in and `other_programs[i][j]`.
-
     """
+
+    def grad(dy):
+        """Calculate the gradients of this inner_product op.
+
+        For empty symbols, this function will just output `None` as the default
+        behavior of tensorflow gradient. For example, no matter what shape is,
+        gradients of a given tensor `x` w.r.t empty `symbol` is None.
+
+
+        >>> qubits = cirq.GridQubit.rect(1, 2)
+        >>> programs = [
+        ...     cirq.Circuit(cirq.H.on_each(qubits)),
+        ...     cirq.Circuit(
+        ...         cirq.X(qubits[0]),
+        ...         cirq.Y(qubits[1]))
+        ... ]
+        >>> other_programs = [
+        ...     cirq.Circuit(cirq.X.on_each(qubits)),
+        ...     cirq.Circuit((cirq.Y**0.125).on_each(qubits)),
+        ...     cirq.Circuit((cirq.X**0.5).on_each(qubits))
+        ... ]
+        >>> programs = tfq.constant([])
+        >>> symbol_names = tf.constant([])
+        >>> symbol_values = tf.constant([])
+        >>> with tf.GradientTape() as t:
+        >>>   t.watch(symbol)
+        >>>   x = tf.constant(tf.ones((3,4)))
+        >>>   y = tfq.math.inner_product(programs, symbol_names,
+        ...                                  symbol_values, other_programs)
+        >>> t.gradient(x, s)  # TensorFlow default behavior.
+        None
+        >>> t.gradient(y, s)
+        None
+
+
+
+        Args:
+            dy: `tf.Tensor` of gradients coming from the next computational op
+                with the shape [batch_size, n_others]
+
+        Returns:
+            `tf.Tensor` with shape [batch_size, n_others, n_symbols] where
+            `out[i][j][k]` is equal to the gradient of the above inner product
+             w.r.t. `symbol_names[k]` merged with the gradient `dy` from the
+             next computational op.
+        """
+        if symbol_names.shape[0] == 0:
+            return [None, None, None, None]
+        inner_prod_grad = inner_product_adj_grad(
+            programs, symbol_names, tf.cast(symbol_values, tf.float32),
+            other_programs)
+        return [None, None, tf.einsum("bos,bo->bos", inner_prod_grad, dy), None]
+
     return MATH_OP_MODULE.tfq_inner_product(programs, symbol_names,
                                             tf.cast(symbol_values, tf.float32),
-                                            other_programs)
+                                            other_programs), grad
```

Comment:
Normally we don't need to write the docstring + function for gradient functionality and we can just decorate the forward pass method with the `@RegisterGradient` decorator. See: https://www.tensorflow.org/api_docs/python/tf/RegisterGradient

Reply:
Let me try it...

Reply:
Hi Mike, I've tried `tf.RegisterGradient` instead of `tf.custom_gradient`, but it seems not supporting complex numbers.

```python
@tf.RegisterGradient("TfqInnerProduct")
def _tfq_inner_product_grad(op, grad):
    programs, symbol_names, symbol_values, other_programs = op.inputs
    if symbol_names.shape[0] == 0:
        return [None, None, None, None]
    inner_prod_grad = inner_product_adj_grad(
        programs, symbol_names, symbol_values, other_programs, grad)
    return [None, None, inner_prod_grad, None]
```

For example, the final result (x) was automatically converted into float, and lost their imaginary parts:
```
Mismatched elements: 28 / 30 (93.3%)
Max absolute difference: 1.05809069
Max relative difference: 0.99853275
 x: array([[ 4.114758e-03,  4.425383e-01, -7.175841e-02],
       [ 8.077559e-02,  1.215895e-02,  1.306026e-01],
       [ 5.921935e-01,  7.048310e-01,  1.323049e-09],...
 y: array([[ 4.109432e-03+7.442631e-03j,  4.425324e-01-7.890046e-01j,
        -7.175633e-02-2.537872e-01j],
       [ 8.077151e-02+4.973777e-01j,  1.216752e-02+2.243689e-01j,...
```

Some investigations said it's intended... [link1](https://github.com/tensorflow/tensorflow/issues/14081), [link2](https://github.com/tensorflow/tensorflow/issues/3346).

So I want to keep this `custom_gradient` implementation. WDYT?

Reply:
Ahhhh ok, thank you for checking. We can keep things using `custom_gradient`. I'll go ahead and review now.